lib/net.c

```diff
@@ -154,120 +154,196 @@ ethertype_filter_add(struct gatekeeper_if *iface, uint16_t ether_type,
 }
 
 /*
- * @dst_ip, @src_port, @src_port_mask, @dst_port, and @dst_port_mask
- * must be in big endian.
- *
- * By specifying the tuple (proto, src_port, dst_port) (and masks),
- * it can filter both IPv4 and IPv6 addresses.
+ * @dst_ip_be, @src_port_be, @src_port_mask_be, @dst_port_be, and
+ * @dst_port_mask_be must be in big endian.
  */
-int
-ntuple_filter_add(struct gatekeeper_if *iface, uint32_t dst_ip,
-	uint16_t src_port, uint16_t src_port_mask,
-	uint16_t dst_port, uint16_t dst_port_mask,
-	uint8_t proto, uint16_t queue_id,
-	int ipv4_configured, int ipv6_configured)
+static int
+ipv4_ntuple_filter_add(struct gatekeeper_if *iface, rte_be32_t dst_ip_be,
+	rte_be16_t src_port_be, rte_be16_t src_port_mask_be,
+	rte_be16_t dst_port_be, rte_be16_t dst_port_mask_be,
+	uint8_t proto, uint16_t queue_id)
 {
-	int ret = 0;
 	unsigned int i;
 	struct rte_eth_ntuple_filter filter_v4 = {
 		.flags = RTE_5TUPLE_FLAGS,
-		.dst_ip = dst_ip,
+		.dst_ip = dst_ip_be,
 		.dst_ip_mask = UINT32_MAX,
 		.src_ip = 0,
 		.src_ip_mask = 0,
-		.dst_port = dst_port,
-		.dst_port_mask = dst_port_mask,
-		.src_port = src_port,
-		.src_port_mask = src_port_mask,
-		.proto = proto,
-		.proto_mask = UINT8_MAX,
-		.tcp_flags = 0,
-		.priority = 1,
-		.queue = queue_id,
-	};
-
-	struct rte_eth_ntuple_filter filter_v6 = {
-		.flags = RTE_5TUPLE_FLAGS,
-		.dst_ip = 0,
-		.dst_ip_mask = 0,
-		.src_ip = 0,
-		.src_ip_mask = 0,
-		.dst_port = dst_port,
-		.dst_port_mask = dst_port_mask,
-		.src_port = src_port,
-		.src_port_mask = src_port_mask,
+		.dst_port = dst_port_be,
+		.dst_port_mask = dst_port_mask_be,
+		.src_port = src_port_be,
+		.src_port_mask = src_port_mask_be,
 		.proto = proto,
 		.proto_mask = UINT8_MAX,
 		.tcp_flags = 0,
 		.priority = 1,
 		.queue = queue_id,
 	};
 
-	RTE_VERIFY(iface->hw_filter_ntuple);
-
-	if (!ipv4_configured)
-		goto ipv6;
-
 	for (i = 0; i < iface->num_ports; i++) {
 		uint16_t port_id = iface->ports[i];
-		ret = rte_eth_dev_filter_ctrl(port_id,
+		int ret = rte_eth_dev_filter_ctrl(port_id,
 			RTE_ETH_FILTER_NTUPLE,
 			RTE_ETH_FILTER_ADD,
 			&filter_v4);
 		if (ret == -ENOTSUP) {
 			G_LOG(ERR,
 				"net: hardware doesn't support adding an IPv4 ntuple filter on port %hhu\n",
 				port_id);
-			ret = -1;
-			goto out;
+			return -1;
 		} else if (ret == -ENODEV) {
```

Comment:
Style: this `if` statement should really be a `switch..case`.

Reply:
I'm OK with this, but note that since the last case is an inequality (`!= 0`), we need an if statement in here somewhere (something like `default: if (ret != 0)...`. Is that what you intended?

Reply:
Good point. We can avoid that `if` statement by adding the following `case` at the very beginning of the `switch..case`:
```
case 0:
	break;
```

Reply:
By the way, log the returned error and `strerror()`; this extra information will be quite helpful if we come across this error in practice.